File: geret/sample/santa_fe_ant_trail/ant.rb

class Ant

  Food = '*'
  Empty = '.'

  Left = { :north => :west, :west => :south, :south => :east, :east => :north }
  Right = { :north => :east, :east => :south, :south => :west, :west => :north }
  DirX = { :north => 0, :west => -1, :south => 0, :east => 1 } 
  DirY = { :north => -1, :west => 0, :south => 1, :east => 0 }

  def initialize
    @grid = []
    IO.read( 'trail.txt' ).each { |line| @grid << line.split( // ) }
    @grid_height = @grid.size
    @grid_width = @grid.max { |line| line.size }

    @dir = :south
    @x, @y = 0, 0
    @consumed_food = 0
  end

  attr_reader :consumed_food, :x, :y, :dir

  def move
    @x, @y = ahead_x, ahead_y  
    next unless @grid[@y][@x] == Food
    @consumed_food += 1
    @grid[@y][@x] = Empty
  end

  def right
    @dir = Right[ @dir ]
  end

  def left
    @dir = Left[ @dir ]
  end

  def food_ahead
    Food == @grid[ ahead_y ][ ahead_x ]
  end

  protected

  def ahead_x
    ( @x + DirX[@dir] ).divmod( @grid_width ).last
  end

  def ahead_y
    ( @y + DirY[@dir] ).divmod( @grid_height ).last
  end

end

File: geret/sample/santa_fe_ant_trail/ant.c
/*
 * This is the Ant library written in the C language
 */

#include <stdio.h>

/*
 * Constants
 */

const int TRUE=1;
const int FALSE=0;
typedef int BOOL;

const char Food = '*';
const char Empty = '.';
const int MaxSteps = 615;

enum Dir {
  North = 0,
  West,
  South,
  East
};

const char Left[4] = { West, South, East, North };
const char Right[4] = { East, North, West, South };
const int DirX[4] = { 0, -1, 0, 1 };
const int DirY[4] = { -1, 0, 1, 0 };
const char Avatar[4] = { '^', '<', 'v', '>' };

/*
 * Playing Field
 */

char * grid[] = {
  ".***............................",
  "...*............................",
  "...*.....................***....",
  "...*....................*....*..",
  "...*....................*....*..",
  "...****.*****........**.........",
  "............*................*..",
  "............*.......*...........",
  "............*.......*...........",
  "............*.......*........*..",
  "....................*...........",
  "............*...................",
  "............*................*..",
  "............*.......*...........",
  "............*.......*.....***...",
  ".................*.....*........",
  "................................",
  "............*...................",
  "............*...*.......*.......",
  "............*...*..........*....",
  "............*...*...............",
  "............*...*...............",
  "............*.............*.....",
  "............*..........*........",
  "...**..*****....*...............",
  ".*..............*...............",
  ".*..............*...............",
  ".*......*******.................",
  ".*.....*........................",
  ".......*........................",
  "..****..........................",
  "................................"
};
const int GridHeight = 32;
const int GridWidth = 32;

/* 
 * Ant State
 */

enum Dir dir = East;
int x = 0;
int y = 0;
int consumed_food = 0;
int steps = 0;

/*
 * Ant Methods
 */

void move()
{
  if( steps >= MaxSteps )
    return;
  x = ahead_x();
  y = ahead_y();
  steps++;
  if( grid[y][x] == Food ) {
    consumed_food++;
    grid[y][x] = Empty;
  }
}

void right()
{
  if( steps >= MaxSteps )
    return;
  dir = Right[ dir ];
  steps++;
}

void left()
{
  if( steps >= MaxSteps )
    return;
  dir = Left[ dir ];
  steps++;
}

BOOL food_ahead()
{
  return Food == grid[ ahead_y() ][ ahead_x() ] ? TRUE : FALSE;
}

void show_scene()
{
  int sx, sy;
  for( sy=0; sy<GridHeight; sy++ ) {
    for( sx=0; sx<GridWidth; sx++ ) {
      printf( "%c", ( x == sx && y == sy ) ? Avatar[dir] : grid[sy][sx] );
    }
    printf("\n");
  }
}

int ahead_x()
{
  return ( x + DirX[dir] ) % GridWidth;
}

int ahead_y()
{
  return ( y + DirY[dir] ) % GridHeight; 
}


